WordPress/gutenberg
# InputControl: Add padding wrapper for prefix/suffix

Follow-up to #42166

## What?

Adds a convenience wrapper component `InputControlPrefix`/`InputControlSuffix` that can be used by a consumer to add standard padding that matches the size variant.

### Wrapped prefix when InputControl is at default size

```jsx
<InputControl
  label="Value"
  placeholder="Placeholder"
  prefix={<InputControlPrefix>@</InputControlPrefix>}
  size="default"
/>
```

<img width="141" alt="Wrapped prefix when InputControl is at default size" src="https://user-images.githubusercontent.com/555336/178581626-bf6b6537-7370-4a94-b524-0b98673d6660.png">

### Wrapped prefix when InputControl is at large size

```jsx
<InputControl
  label="Value"
  placeholder="Placeholder"
  prefix={<InputControlPrefix>@</InputControlPrefix>}
  size="__unstable-large"
/>
```

<img width="161" alt="Wrapped prefix when InputControl is at large size" src="https://user-images.githubusercontent.com/555336/178581662-81ccc169-da8d-437f-bd65-8e76df11b75e.png">

(The padding between the prefix and the placeholder here is too wide — this will be addressed in #42166)

## Why?

Currently, whenever a consumer wants to add padding before their `prefix` element, they need to add it themselves. This is not sustainable because they need to know the padding value of the underlying `InputControl`, and that value also changes depending on the size variant.

Here are the alternative approaches I considered, but decided against:

1. Adding the padding by default, and providing a convenience wrapper that breaks out of that padding. This is similar to the `<CardMedia>` approach taken in the [Card component](https://wordpress.github.io/gutenberg/?path=/docs/components-card--default), and the [approach considered](https://github.com/WordPress/gutenberg/pull/41937#discussion_r914804341) for `Dropdown`. Although it would be nice to have a similar construct for `InputControl`, the current implementation is already no padding by default. Also, it's arguably more natural in this case for the wrapper to _add_ padding instead of break out of it, since we're only dealing with one side of the box. (A padding breakout wrapper deals with all four sides of the box.)
2. Checking if the `prefix` is a function, and if it is, call it with a `{ paddingInlineStart }` argument so the consumer can use that value to style their own margin. This seems simple on the surface, but it's just more work for the consumer, and forces them to use an inline style or a CSS-in-JS solution. Also, I think the general direction we're headed toward for these solutions is to use subcomponents like `CardMedia`.

## How?

Provide a thin wrapper component that applies responsive padding through the Context System. This can be used with any InputControl-based component that inherits the `prefix`/`suffix` props.

## Testing Instructions

1. `npm run storybook:dev`
2. Check out the prefix/suffix stories for `InputControl`.
